### DISCUSSION TITLE
multiple loggers single sink

### DISCUSSION THREAD

**Discussion by jwinarske** [gabime/spdlog#2811] [Q&A]
I need to implement two loggers in the same process.  One is on primary thread (no backtrace), the second is on secondary thread with backtrace enabled.  The primary thread would drain the secondary backtrace to same sink.  What's the recommended pattern for implementing this?

**Comment by jwinarske** [comment]
Something like this?  What about mutex on secondary backtrace?

```
  // main thread
  auto primary_logger = spdlog::get("primary");
  auto secondary_logger = primary_logger->clone("second");
  secondary_logger->enable_backtrace(128);
  // pass pointer of secondary_logger to other thread

  // do some work on secondary thread
  secondary_logger->info("hello");

  // main thread -> periodically call
  secondary_logger->dump_backtrace();
```

**Reply by tt4g** [reply]
> Something like this?

Yes.
See also Wiki: https://github.com/gabime/spdlog/wiki/2.-Creating-loggers#creating-loggers-with-multiple-sinks

> What about mutex on secondary backtrace?

backtracer is thread-safe.

Enable/Disable backtrace (mutex):
  - logger: https://github.com/gabime/spdlog/blob/7e635fca68d014934b4af8a1cf874f63989352b7/include/spdlog/logger-inl.h#L107-L117
  - backtracer: https://github.com/gabime/spdlog/blob/7e635fca68d014934b4af8a1cf874f63989352b7/include/spdlog/details/backtracer-inl.h#L33-L49

Dump backtrace (mutex):
  - logger: https://github.com/gabime/spdlog/blob/7e635fca68d014934b4af8a1cf874f63989352b7/include/spdlog/logger-inl.h#L210-L219
  - backtracer: https://github.com/gabime/spdlog/blob/7e635fca68d014934b4af8a1cf874f63989352b7/include/spdlog/details/backtracer-inl.h#L63-L73

**Reply by jwinarske** [reply]
thanks!

**Reply by jwinarske** [reply]
hmm.  I'm seeing redundant log output even though I'm only writing to secondary logger.

```
auto console_sink = std::make_shared<spdlog::sinks::stdout_color_sink_mt>();
auto logger = std::make_shared<spdlog::logger>("primary", console_sink);
spdlog::set_default_logger(logger);
```

```
auto primary_logger = spdlog::get("primary");
m_vm_logger = primary_logger->clone("vm");
m_vm_logger->enable_backtrace(256);
```
Write Thread
```
m_vm_logger->info(message);
```

Drain Thread
```
m_vm_logger->dump_backtrace();
```

Log has:
```
[info] The Dart VM service is listening on http://127.0.0.1:41845/wQIuWqx9KkQ=/
[info] ****************** Backtrace Start ******************
[info] The Dart VM service is listening on http://127.0.0.1:41845/wQIuWqx9KkQ=/
[info] ****************** Backtrace End ********************
```

I would not expect the leading log entry, only the backtrace dump.

Also is there an option to disable the Backtrace Start/End banner?

**Reply by jwinarske** [reply]
Creating a logger without using clone also produces double log output.  Am I missing something, or is that by design?

**Reply by tt4g** [reply]
This is design. The backtrace API is the ability to replay log messages for debugging.
It is not designed for the purpose of saving logs and outputting them later.

Create a custom sink to buffer logs using ringbuffer_sink (#2774, #1623).
See: https://github.com/gabime/spdlog/wiki/4.-Sinks#implementing-your-own-sink

```cpp
#include <mutex>

#include "spdlog/sinks/base_sink.h"
#include "spdlog/sinks/ringbuffer_sink.h"
#include "spdlog/details/null_mutex.h"

template<typename Mutex>
class buffer_sink : public spdlog::sinks::base_sink <Mutex>
{

...
protected:
    void sink_it_(const spdlog::details::log_msg& msg) override
    {
		this->ringbuffer.log(msg);
    }

    void flush_() override 
    {
        // Impl
    }

private:
	spdlog::sinks::ringbuffer_sink_st ringbuffer;
};

using buffer_sink_mt = buffer_sink<std::mutex>;
using buffer_sink_st = buffer_sink<spdlog::details::null_mutex>;
```

**Comment by gabime** [comment]
If you don't want to see duplicates just set the secondary logger level to  `off`.

**Reply by jwinarske** [reply]
Turning the secondary logger level off removed the duplicates.

@tt4g In the case of ringbuffer is there an easy way to run that through a non-backtrace logger on a lower priority thread without additional malloc/mem copies?

**Reply by tt4g** [reply]
`malloc()` cannot be avoided. Because `ringbuffer_sink::last_formatted()` returns `std::vector`.

https://github.com/gabime/spdlog/blob/7e635fca68d014934b4af8a1cf874f63989352b7/include/spdlog/sinks/ringbuffer_sink.h#L28-L56

If you want to reduce `malloc()`, use `spdlog::details::circular_q` (the container used by ringbuffer_sink).

**Reply by jwinarske** [reply]
which is why I was liking the backtrace implementation.  Only issue is I need to fork it to remove the banners.

**Reply by tt4g** [reply]
> Only issue is I need to fork it to remove the banners.

This can be solved by ringbuffer_sink (#1623), so the PR is closed (#2778).

**Reply by jwinarske** [reply]
I tried out ringbuffer_sink.  After some fiddling I moved to the below pattern.  Simple and works great.

definition
```
auto m_vm_queue = std::make_shared<std::queue<std::string>>();
```
writer thread
```
std::scoped_lock<std::mutex> lock(ctx->m_queue_lock);
ctx->m_vm_queue->push(message);
```
logging thread
```
  if (!m_vm_queue->empty()) {
    std::scoped_lock<std::mutex> lock(m_queue_lock);
    for (int i = 0; i < kVmLogChunkMax; i++) {
      spdlog::info(m_vm_queue->front());
      m_vm_queue->pop();
      if (m_vm_queue->empty())
        break;
    }
  }
```

**Reply by gabime** [reply]
> std::scoped_lock<std::mutex> lock(ctx->m_queue_lock);
ctx->m_vm_queue->push(message);

Lots of allocations with this way..
The ring buffer allocates much less I believe (e.g. if the log message < 250 bytes, it reuses existing pre allocated memory). All that is left is to add some way to iterate it instead of creating a vector(e.g https://github.com/gabime/spdlog/blob/76dfc7e7c0d3c69d3cdaa3399b63545235ccbb02/include/spdlog/details/backtracer-inl.h#L64)

**Reply by jwinarske** [reply]
Are you suggesting to use ringbuffer sink and backtracer?

What's the pattern to remove items from ringbuffer sink?

**Comment by gabime** [comment]
Either use the backtracer, or use the ringbuffer and add a foreach_pop impl (see above) , or maybe even better, iterator support(pr is welcome).

That being said, if the rate of your logs is not high, maybe your solution is good enough.

**Reply by jwinarske** [reply]
I'll take a look at your suggestion.  Thanks.
